File: src/ergmm_latent_effects.c
/*  File src/ergmm_latent_effects.c in package latentnet, part of the
 *  Statnet suite of packages for network analysis, https://statnet.org .
 *
 *  This software is distributed under the GPL-3 license.  It is free,
 *  open source, and has the attribution requirements (GPL Section 7) at
 *  https://statnet.org/attribution .
 *
 *  Copyright 2003-2022 Statnet Commons
 */
/********************************************************/
/* Types of latent space effects supported by latentnet */
/********************************************************/
#include "ergmm_latent_effects.h"
#include <math.h>
/* ERGMM_MCMC_latent_eff = effect of the latent space terms on the linear component. */

/* Define "lookup tables" for latent space effects. */

ERGMM_MCMC_latent_eff_t ERGMM_MCMC_latent_eff[N_LATENT_EFF]={
  ERGMM_MCMC_latent_eff_negative_Euclidean_distance,
  ERGMM_MCMC_latent_eff_dot_product,
  ERGMM_MCMC_latent_eff_negative_Euclidean_distance2
};

/* 
   0 Negative Euclidean distance
*/ 
double ERGMM_MCMC_latent_eff_negative_Euclidean_distance(double *u, double *v, unsigned int dim){
  unsigned int k;
  double dist,dist2=0;
  for(k=0;k<dim;k++){
    dist=u[k]-v[k];
    dist2+=dist*dist;
  }
  return(-sqrt(dist2));
}


/*
  1 Dot product (bilinear) effect
*/
double ERGMM_MCMC_latent_eff_dot_product(double *u, double *v, unsigned int dim){
  double prod=0;
  for(unsigned int i=0; i<dim; i++)
    prod += u[i]*v[i];
  return(prod);
}

/* 
   2 Negative Euclidean distance squared
*/ 
double ERGMM_MCMC_latent_eff_negative_Euclidean_distance2(double *u, double *v, unsigned int dim){
  unsigned int k;
  double dist,dist2=0;
  for(k=0;k<dim;k++){
    dist=u[k]-v[k];
    dist2+=dist*dist;
  }
  return(-dist2);
}

File: R/ergmm.latent.effects.R
#  File R/ergmm.latent.effects.R in package latentnet, part of the Statnet suite
#  of packages for network analysis, http://statnet.org .
#
#  This software is distributed under the GPL-3 license.  It is free,
#  open source, and has the attribution requirements (GPL Section 7) at
#  http://statnet.org/attribution
#
#  Copyright 2003-2017 Statnet Commons
#######################################################################
### effect-specific functions

## negative Euclidean distance
#' @importFrom stats dist
latent.effect.negative.Euclidean<-function(Z){
  -as.matrix(dist(Z))
}

dlpY.dZ.negative.Euclidean<-function(Z,dlpY.deta){
  n<-dim(Z)[1]
  d<-dim(Z)[2]
  Z.invdist<- as.matrix(dist(Z))
  Z.invdist[Z.invdist==0]<-Inf
  Z.invdist<-1/Z.invdist
  dlpY.dZ<-matrix(0,n,d)
  for(k in 1:d){
    Z.normdiff.k<-sapply(1:n,function(j)
                         sapply(1:n,function(i)
                                Z[i,k]-Z[j,k]))*Z.invdist
    dlpY.dZ[,k]<-dlpY.dZ[,k]+
      -sapply(1:n,function(i) crossprod(Z.normdiff.k[i,],dlpY.deta[i,]+dlpY.deta[,i]))
  }
  dlpY.dZ
}


## negative squared Euclidean distance
latent.effect.negative.Euclidean2<-function(Z){
  -as.matrix(dist(Z))^2
}

dlpY.dZ.negative.Euclidean2<-function(Z,dlpY.deta){
  n<-dim(Z)[1]
  d<-dim(Z)[2]
  dlpY.dZ<-matrix(0,n,d)
  for(k in 1:d){
    Z.norm.k<-sapply(1:n,function(j)
                     sapply(1:n,function(i)
                            Z[i,k]-Z[j,k]))*2
    dlpY.dZ[,k]<-dlpY.dZ[,k]+
      -sapply(1:n,function(i) crossprod(Z.norm.k[i,],dlpY.deta[i,]+dlpY.deta[,i]))
  }
  dlpY.dZ
}

## dot product (bilinear)
latent.effect.bilinear<-function(Z){
  tcrossprod(Z)
}

dlpY.dZ.bilinear<-function(Z,dlpY.deta){
  n<-dim(Z)[1]
  d<-dim(Z)[2]
  dlpY.dZ<-matrix(0,n,d)
  for(k in 1:d){
    Z.other<-sapply(1:n,function(j)
                         sapply(1:n,function(i)
                                Z[j,k]))
    dlpY.dZ[,k]<-dlpY.dZ[,k]+
      sapply(1:n,function(i) crossprod(Z.other[i,],dlpY.deta[i,]+dlpY.deta[,i]))
  }
  dlpY.dZ
}

## Dispatcher and information functions

latent.effect.IDs<-list(negative.Euclidean=1,
                        bilinear=2,
                        negative.Euclidean2=3)

latent.effect.names<-c("negative.Euclidean",
                       "bilinear",
                       "negative.Euclidean2")

latent.effect.invariances <- list(negative.Euclidean = c("translation", "rotation", "reflection"),
                                  bilinear = c("rotation", "reflection"),
                                  negative.Euclidean = c("translation", "rotation", "reflection"))


latent.effect.fs<-c(latent.effect.negative.Euclidean,
                    latent.effect.bilinear,
                    latent.effect.negative.Euclidean2)

dlpY.dZ.fs<-c(dlpY.dZ.negative.Euclidean,
              dlpY.dZ.bilinear,
              dlpY.dZ.negative.Euclidean2)
